From f9abccc5ed566fb6397fa50bac0a9eb3ea428026 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:40:05 -0400
Subject: [ethernet] i40e: never generate both software and hardware timestamps

Message-id: <1409658042-22689-245-git-send-email-sassmann@redhat.com>
Patchwork-id: 90437
O-Subject: [RHEL7.1 PATCH 244/281] i40e: never generate both software and hardware timestamps
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jakub Kicinski <kubakici@wp.pl>

skb_tx_timestamp() does not report software time stamp
if SKBTX_IN_PROGRESS is set. According to timestamping.txt
software time stamps are a fallback and should not be
generated if hardware time stamp is provided.

Move call to skb_tx_timestamp() after setting
SKBTX_IN_PROGRESS.

Signed-off-by: Jakub Kicinski <kubakici@wp.pl>
Tested-by: Jim Young <jamesx.m.young@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 259afec7c34bbaa0236edd7e6ad811c4dce2fb1b)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_txrx.c b/drivers/net/ethernet/intel/i40e/i40e_txrx.c
index 2c686e2..f65d361 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_txrx.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_txrx.c
@@ -2285,13 +2285,13 @@ static netdev_tx_t i40e_xmit_frame_ring(struct sk_buff *skb,
  else if (tso)
   tx_flags |= I40E_TX_FLAGS_TSO;
 
- skb_tx_timestamp(skb);
-
  tsyn = i40e_tsyn(tx_ring, skb, tx_flags, &cd_type_cmd_tso_mss);
 
  if (tsyn)
   tx_flags |= I40E_TX_FLAGS_TSYN;
 
+ skb_tx_timestamp(skb);
+
  /* always enable CRC insertion offload */
  td_cmd |= I40E_TX_DESC_CMD_ICRC;
 
-- 
1.7.1